kroxylicious-authorizer-providers/kroxylicious-authorizer-acl/src/main/java/io/kroxylicious/authorizer/provider/acl/AclAuthorizer.java

```diff
@@ -0,0 +1,466 @@
+            if (grant != null) {
+                foundDecision = getDecision(action, grant, whenFound);
+                if (foundDecision != null) {
+                    return foundDecision;
+                }
+            }
+            grant = allowPerPrincipal.lookup(p.getClass(), TypeNameMap.Predicate.TYPE_EQUAL_NAME_STARTS_WITH, p.name());
+            if (grant != null) {
+                foundDecision = getDecision(action, grant, whenFound);
+                if (foundDecision != null) {
+                    return foundDecision;
+                }
+            }
+        }
+        return whenNotFound;
+    }
+
+    @Nullable
+    private static Decision getDecision(Action action,
+                                        PrincipalGrants grants,
+                                        Decision whenFound) {
+        Set<? extends ResourceType<?>> operations;
+        var typeNameMap = grants.nameMatches();
+        ResourceType<?> resourceType = action.operation();
+        if (typeNameMap != null) {
+            operations = typeNameMap.lookup(action.resourceTypeClass(),
+                    TypeNameMap.Predicate.TYPE_EQUAL_NAME_EQUAL,
+                    action.resourceName());
+            if (isFound(operations, resourceType)) {
+                return whenFound;
+            }
+            operations = typeNameMap.lookup(action.resourceTypeClass(),
+                    TypeNameMap.Predicate.TYPE_EQUAL_NAME_ANY,
+                    null);
+            if (isFound(operations, resourceType)) {
+                return whenFound;
+            }
+            operations = typeNameMap.lookup(action.resourceTypeClass(),
+                    TypeNameMap.Predicate.TYPE_EQUAL_NAME_STARTS_WITH,
+                    action.resourceName());
+            if (isFound(operations, resourceType)) {
+                return whenFound;
+            }
+        }
+        var patternMatch = grants.patternMatch();
+        if (patternMatch != null) {
+            operations = patternMatch.lookup(action.resourceTypeClass(), action.resourceName());
+            if (isFound(operations, resourceType)) {
+                return whenFound;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Return true iff the given set of operations allow the given action.
+     */
+    private static boolean isFound(@Nullable Set<? extends ResourceType<?>> operations,
+                                   ResourceType<?> op) {
+        return operations != null && operations.contains(op);
+    }
+
+    @Override
+    public CompletionStage<AuthorizeResult> authorize(Subject subject, List<Action> actions) {
+        List<Action> allowedActions = new ArrayList<>();
+        List<Action> deniedActions = new ArrayList<>();
+        for (var action : actions) {
+            @Nullable
+            Decision decision = authorizeInternal(this.denyPerPrincipal, subject, action, Decision.DENY, null);
```

Comment:
On deny beating allow. This makes it hard to express things like "allow Alice access to all topics, except topics named like 'prefix-*', oh but let her access 'prefix-specialcase'".

I think that's fine, just noting it. I think our expectation is that Users should be starting with it locked down and introducing finer grained allowances, rather than immediately trying to flip things upside-down by giving a principal access to everything then restricting that. It's fighting against our idea of how the authorization should be configured.

But then.. I guess we can also anticipate people immediately trying to do what we don't expect :grinning: